[GeoMechanicsApplication] Added (de)serialization of stress state policies

avdg81 opened the pull request:
**📝 Description**

Stress state policies are typically used and manipulated through a pointer to their abstract base class (i.e. `StressStatePolicy`). This PR adds (de)serialization of the concrete stress state policy classes through a pointer to this interface. Note that it requires some changes in Kratos Core, which have been moved to a [separate PR](https://github.com/KratosMultiphysics/Kratos/pull/13153)~~, that needs to be reviewed and approved by CIMNE~~. The current PR builds on top of that. ~~Once the other PR has been merged with `master`, we can take out the duplicated changes from this PR.~~

avdg81 committed:
Added a unit test for (de)serializing a plane strain stress state policy

avdg81 committed:
Added `load` and `save` members to the stress state policies

To make this work, we had to make a minor change to the (de)serialization code of Kratos Core. We've added a compile-time check to see whether a given class is an abstract class. If it is, we shouldn't try to instantiate it. In practice, this is not expected to ever happen.

avdg81 committed:
Removed the stress state policy factory

It was no longer needed, since we make better use of the (de)serialization code offered by the Kratos Core.

avdg81 committed:
Added one more assertion to the unit test

avdg81 committed:
Added a unit test for (de)serialization of the axisymmetric stress state policy

This required us to register the axisymmetric stress state policy.

avdg81 committed:
Added a unit test for (de)serialization of the interface stress state policy

This required us to register the interface stress state policy.

avdg81 committed:
Added a unit test for (de)serialization of the 3D stress state policy

This required us to register the 3D stress state policy.

avdg81 committed:
Made `#include`s in unit test files more consistent

avdg81 committed:
Improved (de)serialization of `Kratos::shared_ptr`

When the data type of the shared pointer is an abstract base class, no longer attempt to instantiate it (which yields a compile-time error). This would only be attempted when the pointer type is `SP_BASE_CLASS_POINTER`. In practice, the underlying type will always be different from that abstract base class, and therefore the pointer type must always be `SP_DERIVED_CLASS_POINTER`, which looks up the underlying concrete type to instantiate.

avdg81 committed:
Added a unit test for (de)serializing `Kratos::unique_ptr` instances

avdg81 committed:
Improved (de)serialization of `Kratos::intrusive_ptr`

When the data type of the intrusive pointer is an abstract base class, no longer attempt to instantiate it (which yields a compile-time error). This would only be attempted when the pointer type is `SP_BASE_CLASS_POINTER`. In practice, the underlying type will always be different from that abstract base class, and therefore the pointer type must always be `SP_DERIVED_CLASS_POINTER`, which looks up the underlying concrete type to instantiate.

To make this work with the test classes, a few more members had to be defined.

avdg81 committed:
Improved (de)serialization of raw owning pointers

When the data type of the raw owning pointer is an abstract base class, no longer attempt to instantiate it (which yields a compile-time error). This would only be attempted when the pointer type is `SP_BASE_CLASS_POINTER`. In practice, the underlying type will always be different from that abstract base class, and therefore the pointer type must always be `SP_DERIVED_CLASS_POINTER`, which looks up the underlying concrete type to instantiate.

Also changed the test suite of the new unit tests to the one without a kernel, since it's not needed.

avdg81 committed:
Added missing `friend` declarations

avdg81 committed:
Temporarily disable new unit tests in Kratos Core

avdg81 committed:
Attempt to restore new unit test in Kratos Core

Moved the new test classes and the scoped registration class to an anonymous namespace. Need to see whether GCC and Clang then build without errors.

avdg81 committed:
Next attempt to try to understand the build failure

Use a `Serializer` instance rather than a `StreamSerializer` instance.

avdg81 committed:
Next attempt to try to understand the build failure

Made the `Serializer` class a `friend` of the derived test class.

avdg81 committed:
Corrected the `friend` declarations

avdg81 committed:
Copied changes from the other branch

The branch that contains the required changes for Kratos Core has been merged into this branch by copying two files from there to here. This should fix the build problems with GCC and Clang.

avdg81 committed:
Use `StreamSerializer` in the unit tests

markelov208 commented on applications/GeoMechanicsApplication/custom_elements/axisymmetric_stress_state.cpp:
```diff
@@ -71,4 +72,14 @@ SizeType AxisymmetricStressState::GetVoigtSize() const { return GetVoigtSize2D()
 
 SizeType AxisymmetricStressState::GetStressTensorSize() const { return GetStressTensorSize2D(); }
 
+void AxisymmetricStressState::save(Serializer&) const
+{
+    // No data members to be saved (yet)
+}
+
+void AxisymmetricStressState::load(Serializer&)
+{
+    // No data members to be loaded (yet)
+}
+
```
Currently, there are no any data members. Would it be better to have these functions as protected in the base class?

avdg81 replied:
I'm not sure whether I get your point. Both member functions are currently `private`. Can you help me to see why you think they should be `protected`?

markelov208 replied:
Currently, these empty functions shall be implemented in each child. As far as I understand, if the base class defines them as `protected `then all children will use them.

avdg81 replied:
Now I start to see your point. Since the `StressStatePolicy` class is intended to provide an interface only and not an implementation, I'd rather keep the `save` and `load` members pure virtual. In that case, any derived class must implement them, even if they are empty. Since I value this separation of interface and implementation very much, I wouldn't want to change that, unless you have good reasons why we should. Thanks.

markelov208 reviewed (dismissed):
Hi Anne, that is another nice step to have serialization for all classes. I do not have a blocking comment.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/stress_state_policy.h:
```diff
@@ -61,6 +63,12 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) StressStatePolicy
     }
+
+    friend class Serializer;
+
```
Here there are some redundant blank lines

avdg81 replied:
Good catch. Removed the redundant blank lines.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_elements/test_axisymmetric_stress_state.cpp:
```diff
@@ -119,4 +125,24 @@ KRATOS_TEST_CASE_IN_SUITE(AxisymmetricStressState_GivesCorrectStressTensorSize,
+KRATOS_TEST_CASE_IN_SUITE(AxisymmetricStressState_CanBeSavedAndLoadedThroughInterface, KratosGeoMechanicsFastSuiteWithoutKernel)
+{
+    // Arrange
+    RegistrationUtilities::RegisterStressStatePolicies();
```
Should we deregister at the end?

avdg81 replied:
Yes, I agree that it's better to keep the scope of the registration as short as possible (i.e. as long as the test runs, but no longer). I have added an RAII class (`ScopedSerializerRegistrationOfAllStressStatePolicies`) in `test_utilties.h` that takes care of that.

rfaasse reviewed (dismissed):
Nicely done, I don't have any blocking comments, just a minor question and suggestion.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/13136-serialization-of-stress-state-policies

avdg81 committed:
Processed Richard's review suggestions

- Extended class `RegistrationUtilities` with a member function to deregister all stress state policies.
- Added a class for scoped registration of all stress state policies. The unit tests use this class to keep the registration scope as small as possible.
- Removed a few redundant blank lines.
- Construct `std::string` instances directly, rather than through temporary `const char*` objects.

markelov208 reviewed (approved):
Hi Anne, from my viewpoint it is good to be merged.

rfaasse reviewed (approved):
Nice, looks good to go to me!
